wagtail/tests/utils/form_data.py

```diff
@@ -0,0 +1,136 @@
+"""
+The ``assertCanCreate`` method requires page data to be passed in
+the same format that the page edit form would submit. For complex
+page types, it can be difficult to construct this data structure by hand;
+the ``wagtail.tests.utils.form_data`` module provides a set of helper
+functions to assist with this.
+"""
+
+from wagtail.admin.rich_text import get_rich_text_editor_widget
+
+
+def _nested_form_data(data):
+    if isinstance(data, dict):
+        items = data.items()
+    elif isinstance(data, list):
+        items = enumerate(data)
+
+    for key, value in items:
+        key = str(key)
+        if isinstance(value, (dict, list)):
+            for child_keys, child_value in _nested_form_data(value):
+                yield [key] + child_keys, child_value
+        else:
+            yield [key], value
+
+
+def nested_form_data(data):
+    """
+    Translates a nested dict structure into a flat form data dict
+    with hyphen-separated keys.
+
+    .. code-block:: python
+
+        nested_form_data({
+            'foo': 'bar',
+            'parent': {
+                'child': 'field',
+            },
+        })
+        # Returns: {'foo': 'bar', 'parent-child': 'field'}
+    """
+    return {'-'.join(key): value for key, value in _nested_form_data(data)}
+
+
+def streamfield(items):
+    """
+    Takes a list of (block_type, value) tuples and turns it in to
+    StreamField form data. Use this within a :func:`nested_form_data`
+    call, with the field name as the key.
+
+    .. code-block:: python
+
+        nested_form_data({'content': streamfield([
+            ('text', 'Hello, world'),
+        ])})
+        # Returns:
+        # {
+        #     'content-count': '1',
+        #     'content-0-type': 'text',
+        #     'content-0-value': 'Hello, world',
+        #     'content-0-order': 0,
+        #     'content-0-deleted': '',
+        # }
+    """
+    def to_block(index, item):
+        block, value = item
+        return {'type': block, 'value': value, 'deleted': '', 'order': index}
+    data_dict = {str(index): to_block(index, item)
+                 for index, item in enumerate(items)}
+    data_dict['count'] = str(len(data_dict))
```

Comment:
Why does `'count'` has to be a string?

Reply:
It probably doesn't, but at least this way we can be sure it exactly matches a real form submission. It looks like the one place we're _not_ converting to string is the `'order'` field above - I'll update this for consistency.

Reply:
Ah, right, makes sense.